src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -0,0 +1,191 @@
+                    continue;
+                }
+
+                List<OutboundMatch> matches;
+                if (!namedOutboundMatches.TryGetValue(entry.RouteName, out matches))
+                {
+                    matches = new List<OutboundMatch>();
+                    namedOutboundMatches.Add(entry.RouteName, matches);
+                }
+                matches.Add(outboundMatch);
+            }
+
+            return (allOutboundMatches, namedOutboundMatches);
+        }
+
+        private OutboundRouteEntry CreateOutboundRouteEntry(MatcherEndpoint endpoint)
+        {
+            var routeNameMetadata = endpoint.Metadata.GetMetadata<IRouteNameMetadata>();
+            var entry = new OutboundRouteEntry()
+            {
+                Handler = NullRouter.Instance,
+                Order = endpoint.Order,
+                Precedence = RoutePrecedence.ComputeOutbound(endpoint.ParsedTemplate),
+                RequiredLinkValues = endpoint.RequiredValues,
+                RouteTemplate = endpoint.ParsedTemplate,
+                Data = endpoint,
+                RouteName = routeNameMetadata?.Name,
+            };
+
+            // TODO: review. These route constriants should be constructed when the endpoint
+            // is built. This way they can be checked for validity on app startup too
+            var constraintBuilder = new RouteConstraintBuilder(
+                _inlineConstraintResolver,
+                endpoint.ParsedTemplate.TemplateText);
+            foreach (var parameter in endpoint.ParsedTemplate.Parameters)
+            {
+                if (parameter.InlineConstraints != null)
+                {
+                    if (parameter.IsOptional)
+                    {
+                        constraintBuilder.SetOptional(parameter.Name);
+                    }
+
+                    foreach (var constraint in parameter.InlineConstraints)
+                    {
+                        constraintBuilder.AddResolvedConstraint(parameter.Name, constraint.Constraint);
+                    }
+                }
+            }
+            entry.Constraints = constraintBuilder.Build();
+            entry.Defaults = new RouteValueDictionary(endpoint.Defaults);
+            return entry;
+        }
+
+        private IDictionary<string, NamedMatch> GetNamedMatches(
+            IDictionary<string, List<OutboundMatch>> namedOutboundMatches)
+        {
+            var result = new Dictionary<string, NamedMatch>(StringComparer.OrdinalIgnoreCase);
+            foreach (var namedOutboundMatch in namedOutboundMatches)
+            {
+                var namedEntry = new NamedMatch();
+                if (namedOutboundMatch.Value.Count == 1)
+                {
+                    namedEntry.Match = namedOutboundMatch.Value[0];
+                }
+                else
+                {
+                    namedEntry.LinkGenerationTree = new LinkGenerationDecisionTree(namedOutboundMatch.Value.ToArray());
+                }
```

Comment:
I don't think we should treat a single endpoint as different - I can't think of a reason why we want these cases to behave differently